patchwork/scheduling.py

```diff
@@ -178,16 +178,20 @@ def resolve_action(self, starting_context: Context, action: Action) -> Optional[
             self.memoizer.forget(starting_context)
             raise
 
-    def choose_arbitrary_context(self) -> Optional[Context]:
-        if len(self.pending_contexts) > 0:
-            choice = self.pending_contexts.popleft()
-            self.active_contexts.add(choice)
-            return choice
-        return None
+    def choose_context_lazily(self) -> Optional[Context]:
+        if not self.pending_contexts:
+            return None
+
+        choice = next((c for c in self.pending_contexts
+                       if c.is_blocking(self.db)),
+                      self.pending_contexts[0])
```

Comment:
When is the default used?

Reply:
After H sends a reply, given there are no blocked contexts left. For example, on every `sess.act(Reply(…))` in `test_basic.TestBasic.testRecursion`. (`testRecursion` is the 351*5019 scenario from the taxonomy, implemented with one level of recursion.) It goes like this:

1. `Reply(a).execute(…)` returns no mandatory successor context. → `resolve_action` returns `None`. → The scheduler has to choose a pending context.
2. The context that was waiting for answer `a` is now unblocked. → No context is blocked anymore. → Default case.

Also, if we supported multi-session usage and a user requested a context to work on, the default case would give her one, even when there are only non-blocking contexts. This is what I mean when I write that the system is not totally lazy. Maybe you would prefer total laziness? My thoughts were: Busy laziness is a means of prioritizing blocking contexts and otherwise causes link prefetching. Total laziness allows for elegant programs such as `(take 10 (fibonacci-numbers))`, but I'd expect it to increase question-answer latency in the real world.

Reply:
I think `choose_context_lazily` should implement "total laziness", but I also think you're right that we should keep around functionality for doing work on non-blocking contexts. I'd do this by making `_choose_next_context` try `choose_context_lazily` and if it doesn't succeed call `choose_arbitrary_context` (which you'd need to reinstate).

That said, once we figure out how to implement `choose_context_to_advance_promise` and can limit work to a particular root question, we should be totally lazy about that and only do work that's actually required for the given question.

Reply:
It didn't make sense to eliminate it, but I pushed a draft of choose_context_to_advance_promise.